src/System.Linq/src/System/Linq/Where.cs

```diff
@@ -271,34 +432,86 @@ public override Iterator<TResult> Clone()
                 return new WhereSelectArrayIterator<TSource, TResult>(_source, _predicate, _selector);
             }
 
-            public override bool MoveNext()
+            public int GetCount(bool onlyIfCheap)
             {
-                if (_state == 1)
+                if (onlyIfCheap)
                 {
-                    while (_index < _source.Length)
+                    return -1;
+                }
+
+                int count = 0;
+
+                foreach (TSource item in _source)
+                {
+                    if (_predicate(item))
                     {
-                        TSource item = _source[_index];
-                        _index++;
-                        if (_predicate(item))
+                        checked
                         {
-                            _current = _selector(item);
-                            return true;
+                            count++;
```

Comment:
We should revisit the change in #11841. I think we may have gone too far. 
It seems ok to use GetCount(onlyIfCheap) to preallocate buffers, but skipping selectors  in a context of aggregating operator (even if trivial - like Count) now seems potentially breaking.

---
In reply to: [88977111](https://github.com/dotnet/corefx/pull/12703#discussion_r88977111) [](ancestors = 88977111,88807476)